src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * A sequence of steps that progressively fill in the ContextShard.
+ *
+ * Initially we have only the interval and variants.
+ * subdivideAndFillReads then loads the reads (and splits into smaller shards)
+ * fillContext then adds the reference bases.
+ *
+ * In the end we have reads, variants, and reference bases.
+ */
+public class AddContextDataToReadOptimized {
+
+    // Immutable.
+    public static class ContextShard implements Serializable {
+        private static final long serialVersionUID = 1L;
+
+        // the interval covered by this shard
+        public final SimpleInterval interval;
+        // variants that overlap with the shard.
+        public final IntervalsSkipList<Variant> variants;
+        // reads that start in the shard
+        public final ArrayList<GATKRead> reads;
+        // variants and reference for the particular read at the same index as this element.
+        public final ArrayList<ReadContextData> readContext;
+
+        public ContextShard(SimpleInterval interval) {
+            this.interval = interval;
+            this.variants = null;
+            this.reads = new ArrayList<>();
+            this.readContext = new ArrayList<>();
+        }
+
+        /**
+         * Careful: this ctor takes ownership of the passed reads and ReadContextData array.
+         * Do not modify them after this call (ideally don't even keep a reference to them).
+         */
+        private ContextShard(SimpleInterval interval, IntervalsSkipList<Variant> variants, final ArrayList<GATKRead> reads, final ArrayList<ReadContextData> readContext) {
+            this.interval = interval;
+            this.variants = variants;
+            this.reads = reads;
+            this.readContext = readContext;
+        }
+
+        /**
+         * create a new shard, keeping only the variants that overlap
+         * with the new interval. Reads etc are unchanged.
+         */
+        public ContextShard split(SimpleInterval newInterval) {
+            final IntervalsSkipList<Variant> newVariants;
+            if (null==variants) {
+                newVariants = null;
+            } else {
+                newVariants = new IntervalsSkipList<>( variants.getOverlapping(newInterval) );
+            }
+            return new ContextShard(newInterval, newVariants, reads, readContext);
+        }
+
+        /**
+         * creates a new shard, adding the specified variants.
+         */
+        public ContextShard withVariants(ArrayList<Variant> newVariants) {
+            return new ContextShard(this.interval, new IntervalsSkipList<>(newVariants), reads, readContext);
```

Comment:
Will the object be in a consistent state after this call? Won't the variants in `readContext` be out of sync with `newVariants`?